mailing/sendgrid.go

```diff
@@ -12,7 +13,15 @@ import (
 //
 //	Reference: https://github.com/sendgrid/sendgrid-go
 type sendgridEmailService struct {
-	client sendgridSender
+	client          sendgridSender
+	onlineTemplates bool
```

Comment:
> Since you've already gone through the trouble of extracting the `emailBuilder`, I believe it makes sense to make this implementation a little more generic and have the email content logic be provided by dependency injection.
> 
> We should change the `onlineTemplates` field from being an option to instead being an object that receives the `emailBuilder`, `template` and `data` and configures `emailBuilder` however it sees fit.
> 
> ```go
> type contentInjector func(emailBuilder, template string, data any) error
> 
> type sendgridEmailService struct {
> 	client          sendgridSender
> 	contentInjector
> }
> 
> func (s *sendgridEmailService) Send([...]) [...] {
> 	[...]
> 	// This replaces the onlineTemplates if statement block
> 	s.contentInjector(builder, template, data)
> 	[...]
> }
> ```
> 
> That way we can define content injectors as needed and pass them over to the implementation.
> 
> ```go
> func htmlContentInjector(builder emailBuilder, template string, data any) error {
> 	htmlContent, err := gz.ParseHTMLTemplate(template, data)
> 	if err != nil {
> 		return err
> 	}
> 	builder.Content("text/html", htmlContent)
> 	return nil
> }
> 
> func templateContentInjector(builder emailBuilder, template string, data any) error {
> 	builder.Template(template, data)
> 	return nil
> }
> ```
> 
> We should then have two different instantiation functions `NewSendgridEmailSender` and `NewSendgridTemplateEmailSender`, which correspond to the respective content injectors.
> 
> ```go
> func NewSendgridEmailSender(client sendgridSender) Sender {
> 	return &sendgridEmailService{
> 		client:          client,
> 		contentInjector: htmlContentInjector,
> 	}
> }
> 
> func NewSendgridTemplateEmailSender(client sendgridSender) Sender {
> 	return &sendgridEmailService{
> 		client:          client,
> 		contentInjector: templateContentInjector,
> 	}
> }
> ```
> 
> Or have `NewSendgridEmailSender` require `contentInjector` and have the previously defined content injectors be public instead of private.
> 
> ```go
> func NewSendgridEmailSender(client sendgridSender, contentInjector ContentInjector) Sender {
> 	return &sendgridEmailService{
> 		client:          client,
> 		contentInjector: contentInjector,
> 	}
> }
> 
> // Elsewhere in the application
> package main
> 
> func main() {
> 	// Assume client was instantiated before this call
> 	mailing.NewSendgridEmailSender(client, mailing.HTMLContentInjector)
> }
> ```

I like this proposal, it looks like the very natural next step for this implementation.